File: leetcode3/남효정/3996. Even Number of Knight Moves.py
# 풀이 실패
# 나이트는 1번 움직일 때마다 무조건 현재 칸과 다른 색의 칸으로 감
# 따라서 짝수 번 움직이면 원래 색으로 돌아감
# x + y 좌표 합의 홀짝으로 칸의 색상을 알 수 있음
class Solution:
    def canReach(self, start: list[int], target: list[int]) -> bool:
        return (start[0] + start[1]) % 2 == (target[0] + target[1]) % 2

File: leetcode3/황은지/3996. Even Number of Knight Moves.js
/**
 * @param {number[]} start
 * @param {number[]} target
 * @return {boolean}
 */
var canReach = function (start, target) {
  const dirs = [
    [-1, -2],
    [-2, -1],
    [-2, 1],
    [-1, 2],
    [1, 2],
    [2, 1],
    [2, -1],
    [1, -2],
  ];
  const visited = Array(8);
  for (let i = 0; i < 8; i++) {
    visited[i] = Array.from({ length: 8 }, () => Array(2));
  }
  const startX = start[0];
  const startY = start[1];
  const queue = [[startX, startY, true]];
  let head = 0;
  visited[startX][startY][1] = true;

  while (queue.length - head > 0) {
    const [curX, curY, curFlag] = queue[head++];
    if (curX === target[0] && curY === target[1] && curFlag) return true;
    for (const [dirX, dirY] of dirs) {
      const nextX = dirX + curX;
      const nextY = dirY + curY;
      const nextFlag = !curFlag;

      if (nextX < 0 || nextY < 0 || nextX >= 8 || nextY >= 8) continue;
      if (visited[nextX][nextY][nextFlag === true ? 1 : 0]) {
        continue;
      }
      visited[nextX][nextY][nextFlag === true ? 1 : 0] = true;
      queue.push([nextX, nextY, nextFlag]);
    }
  }
  return false;
};
